### PR TITLE
Kamu UI 302 integrate hard compacting flow type into flows system UI

### PR DESCRIPTION
**Pull request by dmitriy-borzenko**

### CODE REVIEW

File: src/app/api/gql/compacting-dataset/dataset-flow-compacting.graphql [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            ... on FlowIncompatibleDatasetKind {
+              message
+              expectedDatasetKind
+              actualDatasetKind
+            }
+
+            ... on FlowInvalidBatchingConfig {
+              message
+              reason
+            }
```
Comment (zaychenko-sergei) [inline src/app/api/gql/compacting-dataset/dataset-flow-compacting.graphql:30]: This error is impossible in compacting flow

File: src/app/common/app.values.ts [hunk 2]
```diff
@@ -15,6 +15,7 @@ export default class AppValues {
     public static readonly DISPLAY_DATE_FORMAT = "DD MMM YYYY";
     public static readonly DISPLAY_TOOLTIP_DATE_FORMAT = "MMM D, YYYY, HH:mm A";
     public static readonly CRON_EXPRESSION_DATE_FORMAT = "MMM Do YYYY, h:mm:ss A";
+    public static readonly DISPLAY_FLOW_DATE_FORMAT = "y-MM-dd, h:mm:ss a";
```
Comment (zaychenko-sergei) [inline src/app/common/app.values.ts:18]: Why do we need yet another date format?

File: src/app/components/data-access-panel/data-access-panel.component.scss [hunk 3]
```diff
@@ -10,6 +10,7 @@
     .get-data-text {
         color: #1a7f37;
         font-weight: 700;
+        font-size: 17px !important;
```
Comment (zaychenko-sergei) [inline src/app/components/data-access-panel/data-access-panel.component.scss:13]: I remember Oleksiy was reacting bad on `!important`. Is it strictly necessary?

File: resources/schema.graphql [hunk 4]
```diff
@@ -377,6 +383,10 @@ type DatasetFlowConfigsMut {
     paused: Boolean!
     batching: BatchingConditionInput!
   ): SetFlowConfigResult!
+  setConfigCompacting(
+    datasetFlowType: DatasetFlowType!
+    compactingArgs: CompactingConditionInput!
+  ): SetFlowConfigResult!
```
Comment (zaychenko-sergei) [inline resources/schema.graphql:389]: Suggestion for you and Roma: let's split data types we return from different `setConfig...` methods, as different groups of errors are possible - as a result of current design, the fragments contain impossible variants.

File: src/app/dataset-flow/dataset-flow-details/tabs/flow-details-history-tab/flow-details-history-tab.helpers.ts [hunk 5]
```diff
@@ -80,12 +80,12 @@ export class DatasetFlowDetailsHelpers {
     }
 
     public static flowOutcomeOptions(outcome: MaybeNullOrUndefined<FlowOutcome>): { icon: string; class: string } {
-        switch (outcome) {
-            case FlowOutcome.Success:
+        switch (outcome?.__typename) {
```
Comment (zaychenko-sergei) [inline src/app/dataset-flow/dataset-flow-details/tabs/flow-details-history-tab/flow-details-history-tab.helpers.ts:83]: I think `flowOutcomeOptions` may accept simply `FlowOutcome` without `null` or `undefined`.
The outcome is surely available in `Finished` state only, and you should rather assert it is present at higher level, than here.

File: src/app/dataset-flow/dataset-flow-details/tabs/flow-details-history-tab/flow-details-history-tab.helpers.ts [hunk 6]
```diff
@@ -141,8 +141,20 @@ export class DatasetFlowDetailsHelpers {
                                                       : "blocks"
                                               }`
                                             : "Dataset is up-to-date";
+
+                                    case "FlowDescriptionDatasetHardCompacting":
+                                        switch (flowDetails.description.compactingResult?.__typename) {
+                                            case "FlowDescriptionHardCompactingSuccess":
+                                                return `There were ${flowDetails.description.compactingResult.originalBlocksCount} block(s), now there are ${flowDetails.description.compactingResult.resultingBlocksCount} block(s)`;
```
Comment (zaychenko-sergei) [inline src/app/dataset-flow/dataset-flow-details/tabs/flow-details-history-tab/flow-details-history-tab.helpers.ts:148]: Compacted N original blocks to M resulting blocks

File: src/app/dataset-view/additional-components/dataset-settings-component/dataset-settings.component.html [hunk 7]
```diff
@@ -11,7 +11,9 @@
                                 *ngIf="
                                     navItem.activeTab === settingsTabsEnum.SCHEDULING
                                         ? isSchedulingAvailable
-                                        : navItem.visible
+                                        : navItem.visible ||
+                                          (navItem.activeTab === settingsTabsEnum.COMPACTING &&
+                                              datasetBasics.kind === DatasetKind.Root)
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/dataset-settings-component/dataset-settings.component.html:16]: Move the logic to component and test it better. This is becoming complex.

File: src/app/dataset-view/additional-components/dataset-settings-component/dataset-settings.component.html [hunk 8]
```diff
@@ -47,5 +49,8 @@
                 data-test-id="settings-scheduling-tab"
             ></app-dataset-settings-scheduling-tab>
         </ng-container>
+        <ng-container *ngIf="activeTab === settingsTabsEnum.COMPACTING && datasetBasics.kind === DatasetKind.Root">
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/dataset-settings-component/dataset-settings.component.html:52]: Don't place logic to HTML

File: src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            <div class="row">
+                <div class="col-md-3">
+                    <div class="d-flex align-items-center pt-1">
+                        <app-tooltip-icon class="mt-1" [tooltip]="MAX_SLICE_SIZE_TOOLTIP"></app-tooltip-icon>
+                        <span class="d-inline-block">Max slice size:</span>
+                    </div>
+                </div>
+                <div class="col-md-9 d-flex">
+                    <select class="form-select me-4 form-control-width" formControlName="sliceUnit">
+                        <option [value]="SliceUnit.B">B</option>
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html:17]: Just bytes might be unnecessary from user point of view. Let's have kilobytes as a minimal suggested unit.

File: src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+                        <option [value]="SliceUnit.KB">KB</option>
+                        <option [value]="SliceUnit.MB">MB</option>
+                        <option [value]="SliceUnit.GB">GB</option>
+                    </select>
+                    <input type="number" min="1" class="form-control form-control-width" formControlName="sliceSize" />
+                    <ng-container
+                        *ngIf="
+                            hardCompactionForm?.get('sliceSize')?.errors?.minNumber &&
+                            (hardCompactionForm.get('sliceSize')?.touched || hardCompactionForm.get('sliceSize')?.dirty)
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html:26]: It's usually cleaner to provide a component property that returns a particular form control.

File: src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+                        type="number"
+                        min="1"
+                        class="form-control form-control-width"
+                        formControlName="recordsCount"
+                    />
+                    <ng-container
+                        *ngIf="
+                            hardCompactionForm?.get('recordsCount')?.errors?.minNumber &&
+                            (hardCompactionForm.get('recordsCount')?.touched ||
+                                hardCompactionForm.get('recordsCount')?.dirty)
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html:57]: Similar here

File: src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+                        "
+                    >
+                        <span
+                            class="text-danger ms-4 fs-12 d-flex align-items-end text-break"
+                            data-test-id="slice-size-error"
+                        >
+                            The value must be positive
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html:64]: Repeating text

File: src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.ts [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+@Component({
+    selector: "app-dataset-settings-compacting-tab",
+    templateUrl: "./dataset-settings-compacting-tab.component.html",
+    styleUrls: ["./dataset-settings-compacting-tab.component.scss"],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class DatasetSettingsCompactingTabComponent implements OnInit {
+    @Input() public datasetBasics: DatasetBasicsFragment;
+    public hardCompactionForm = this.fb.group({
+        sliceUnit: [SliceUnit.GB, [Validators.required]],
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.ts:23]: Let's ask SM about the expected defaults. I think 1GB might be a bit too heavyweight slice for transfer protocol parallelism. I think the optimal value is likely in tens of megabytes.

File: src/app/dataset-view/additional-components/flows-component/components/tile-base-widget/tile-base-widget.component.ts [hunk 14]
```diff
@@ -26,4 +25,10 @@ export class TileBaseWidgetComponent {
     public tileWidgetClass(node: FlowSummaryDataFragment): string {
         return TileBaseWidgetHelpers.tileWidgetClass(node);
     }
+
+    public tileOutcomeMessage: Record<string, string> = {
+        FlowSuccessResult: "success",
+        FlowFailedError: "failed",
+        FlowAbortedResult: "aborted",
+    };
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/flows-component/components/tile-base-widget/tile-base-widget.component.ts:33]: Saw similar code somewhere above

File: src/assets/svg/icons.json [hunk 15]
```diff
@@ -127,6 +127,10 @@
     {
       "iconName": "hour-glass",
       "iconPath": "/assets/svg/hour-glass.svg"
+    },
+    {
+      "iconName": "compress",
```
Comment (zaychenko-sergei) [inline src/assets/svg/icons.json:132]: Compacting is not compressing

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]

File: src/app/components/data-access-panel/data-access-panel.component.scss (hunk 3 above)
Comment (dmitriy-borzenko) [reply src/app/components/data-access-panel/data-access-panel.component.scss:13]: Yes, fixed

File: src/app/dataset-flow/dataset-flow-details/tabs/flow-details-history-tab/flow-details-history-tab.helpers.ts (hunk 5 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-flow/dataset-flow-details/tabs/flow-details-history-tab/flow-details-history-tab.helpers.ts:83]: Fixed

File: src/app/dataset-flow/dataset-flow-details/tabs/flow-details-history-tab/flow-details-history-tab.helpers.ts (hunk 6 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-flow/dataset-flow-details/tabs/flow-details-history-tab/flow-details-history-tab.helpers.ts:148]: Changed

File: resources/schema.graphql (hunk 4 above)
Comment (dmitriy-borzenko) [reply resources/schema.graphql:389]: +

File: src/app/dataset-view/additional-components/dataset-settings-component/dataset-settings.component.html (hunk 8 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/dataset-settings-component/dataset-settings.component.html:52]: Moved logic to component

File: src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html (hunk 9 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html:17]: Removed SliceUnit.B

File: src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.ts (hunk 13 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.ts:23]: ### +

File: src/app/dataset-view/additional-components/flows-component/components/tile-base-widget/tile-base-widget.component.ts (hunk 14 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/flows-component/components/tile-base-widget/tile-base-widget.component.ts:33]: Fixed

File: src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html (hunk 11 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html:57]: Fixed

File: src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html (hunk 10 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html:26]: Fixed

File: src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html (hunk 12 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/dataset-settings-component/tabs/compacting/dataset-settings-compacting-tab.component.html:64]: Fixed

File: src/app/dataset-view/additional-components/dataset-settings-component/dataset-settings.component.html (hunk 7 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/dataset-settings-component/dataset-settings.component.html:16]: Moved logic to the component

File: src/app/dataset-view/additional-components/flows-component/components/tile-base-widget/tile-base-widget.component.ts [hunk 16]
```diff
@@ -26,4 +25,23 @@ export class TileBaseWidgetComponent {
     public tileWidgetClass(node: FlowSummaryDataFragment): string {
         return TileBaseWidgetHelpers.tileWidgetClass(node);
     }
+
+    // public tileOutcomeMessage: Record<string, string> = {
+    //     FlowSuccessResult: "success",
+    //     FlowFailedError: "failed",
+    //     FlowAbortedResult: "aborted",
+    // };
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/flows-component/components/tile-base-widget/tile-base-widget.component.ts:33]: garbage

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]
image

I don't see how compacting error for derived datasets is integrated.

File: src/app/dataset-view/additional-components/flows-component/components/tile-base-widget/tile-base-widget.component.ts (hunk 16 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/flows-component/components/tile-base-widget/tile-base-widget.component.ts:33]: Removed

**Review by zaychenko-sergei — APPROVED** [review]

**Pull request merged**